src/scripting/lua_widget_attributes.cpp

```diff
@@ -0,0 +1,527 @@
+	lua_pushvalue(L, value.index);
+	lua_call(L, 2, 0);
+}
+
+WIDGET_SETTER("visible", lua_index_raw, gui2::styled_widget)
+{
+
+	typedef gui2::styled_widget::visibility visibility;
+
+	visibility flag = visibility::visible;
+
+	switch(lua_type(L, value.index)) {
+		case LUA_TBOOLEAN:
+			flag = luaW_toboolean(L, value.index)
+					? visibility::visible
+					: visibility::invisible;
+			break;
+		case LUA_TSTRING:
+			{
+				const std::string& str = lua_tostring(L, value.index);
+				if(str == "visible") {
+					flag = visibility::visible;
+				} else if(str == "hidden") {
+					flag = visibility::hidden;
+				} else if(str == "invisible") {
+					flag = visibility::invisible;
+				} else {
+					luaL_argerror(L, value.index, "string must be one of: visible, hidden, invisible");
+				}
+			}
+			break;
+		default:
+			luaW_type_error(L, value.index, "boolean or string");
+	}
+
+	w.set_visible(flag);
+
+	//if(flag == visibility::hidden) {
+	//	// HACK: this is needed to force the widget to be repainted immediately
+	//	//       to get rid of its ghost image.
+	//	scoped_dialog::current->window->invalidate_layout();
+	//}
+}
+
+//must be last
+WIDGET_SETTER("value_compat,label", t_string, gui2::styled_widget)
+{
+	w.set_label(value);
+}
+
+WIDGET_GETTER("type", std::string, gui2::widget)
+{
+	if(gui2::styled_widget* sw = dynamic_cast<gui2::styled_widget*>(&w)) {
+		return sw->get_control_type();
+	}
+	else if(dynamic_cast<gui2::tree_view_node*>(&w)) {
+		return "tree_view_node";
+	}
+	else if(dynamic_cast<gui2::grid*>(&w)) {
+		return "grid";
+	}
+	else {
+		return "";
+	}
+}
+
+///////////////////////////////////////////////////////
+////////////////////// CALLBACLS //////////////////////
+///////////////////////////////////////////////////////
+namespace {
+
+void dialog_callback(lua_State* L, lua_ptr<gui2::widget>& wp, const std::string& id)
+{
+	gui2::widget* w = wp.get_ptr();
+	if(!w) {
+		ERR_LUA << "widget was deleted\n";
+		return;
+	}
+	gui2::window* wd = w->get_window();
+	if(!wd) {
+		ERR_LUA << "canot find window in eidgte callback\n";
+		return;
+	}
+	luaW_callwidgetcallback(L, w, wd, id);
+}
+
+WIDGET_SETTER("on_modified", lua_index_raw, gui2::widget)
+{
+	gui2::window* wd = w.get_window();
+	if(!wd) {
+		throw std::invalid_argument("the widget has no window assigned");
+	}
+	lua_pushvalue(L, value.index);
+	if (!luaW_setwidgetcallback(L, &w, wd, "on_modified")) {
+		connect_signal_notify_modified(w, std::bind(&dialog_callback, L, lua_ptr<gui2::widget>(w), "callback"));
+	}
+}
+
+WIDGET_SETTER("on_left_click", lua_index_raw, gui2::widget)
+{
+	gui2::window* wd = w.get_window();
+	if(!wd) {
+		throw std::invalid_argument("the widget has no window assigned");
+	}
+	lua_pushvalue(L, value.index);
+	if (!luaW_setwidgetcallback(L, &w, wd, "on_left_click")) {
+		connect_signal_notify_modified(w, std::bind(&dialog_callback, L, lua_ptr<gui2::widget>(w), "callback"));
+	}
+}
+
+WIDGET_SETTER("on_button_click", lua_index_raw, gui2::widget)
```

Comment:
I don't understand the difference between `on_left_click` and `on_button_click`. They sound like they both do the same thing?

Reply:
On left_click is whenever  the user clicks on the widget (this works for widgets) while on button_click only works on widgets and depends on the specific type of button, in particular for "repeating_button" it will fire the event multiepl times if the user keeps pressing the button

Reply:
Hmm. I think I'd recommend renaming them to something like `on_mouse_down` (or `on_mouse_up`, whichever more accurately represents it) for the one that applies to any widget and `on_click` for the one that only applies to buttons.